resource/traversers/dfu_impl.hpp

```diff
@@ -71,7 +71,8 @@ struct jobmeta_t {
         jobid = id;
         alloc_type = alloc;
         int64_t g_duration = std::chrono::duration_cast<std::chrono::seconds>
-            (graph_duration.graph_end - graph_duration.graph_start).count ();
+            (graph_duration.graph_end -
+             std::chrono::system_clock::now()).count ();
```

Comment:
Sorry @grondo, but after more thought my previous suggestion results in a small drift, too. Can you do this instead:
```suggestion
            (graph_duration.graph_end - now);
 ```

Reply:
Will do! However, for the uninitiated, can you describe the problem being fixed by this change? I only understand enough of `std::chrono` to make this one change, and with this subtle of a change it seems this might be fairly brittle. At least I could add a comment here for future code spelunkers.

Reply:
The motivation for my request was to reduce the number of independent measurements of `now` throughout Fluxion which I hope makes drift debugging easier.

Your question is good and caused me to reconsider the `g_duration` computation and the downstream use. I think there's a bigger problem with the logic of aligning the job duration with the `graph_end` time. Issue #1103 reveals that the check here (which was designed to catch the problem) isn't working correctly: https://github.com/flux-framework/flux-sched/blob/bf62d8a74591353cf2aff43e9aeb41be81da8bef/resource/traversers/dfu.cpp#L323

Reply:
I only just realized what you said initially: that I was redeclaring `now` in the original patch. Somehow I missed that there was already a `now` declared in the function. :facepalm: Sorry about that!

However, using the existing `int64_t` now results in a compilation error. I'm guessing some kind of cast needs to be used, but being pretty unfamilar with `std::chrono` I'm not sure what to use.

Since `g_duration` and `now` are already `int64_t` type, would it make sense to just cast `graph_duration.graph_end` to `int64_t` as well before taking the difference, e.g.

```diff
diff --git a/resource/traversers/dfu_impl.hpp b/resource/traversers/dfu_impl.hpp
index 098fcd41..f9fa8f9b 100644
--- a/resource/traversers/dfu_impl.hpp
+++ b/resource/traversers/dfu_impl.hpp
@@ -70,9 +70,9 @@ struct jobmeta_t {
         now = t;
         jobid = id;
         alloc_type = alloc;
-        int64_t g_duration = std::chrono::duration_cast<std::chrono::seconds>
-            (graph_duration.graph_end -
-             std::chrono::system_clock::now()).count ();
+        int64_t g_end = std::chrono::system_clock::to_time_t
+                          (graph_duration.graph_end);
+        int64_t g_duration = g_end - now;
 
         if (g_duration <= 0) {
             errno = EINVAL;

```

I really know nothing about `std::chrono` and was honestly having trouble casting `now` to the appropriate type otherwise. Please feel free to suggest anything you'd like (or even go ahead and  push the correct fix to this branch).

Reply:
> Your question is good and caused me to reconsider the g_duration computation and the downstream use. I think there's a bigger problem with the logic of aligning the job duration with the graph_end time. Issue https://github.com/flux-framework/flux-sched/issues/1103 reveals that the check here (which was designed to catch the problem) isn't working correctly:

Ah, yes, I see now that the "fix" attempted in this PR won't really fix anything. If the start of the job is delayed, then its expiration may still be greater than the instance expiration because real time has passed between when `build()` is called and the job's actual starttime.

I think the real fix is in the code you've called out above, or to have a way to just set the job _expiration_ to the instance expiration, regardless of the job's duration (though I don't know the code well enough to determine how to do that).

Also, unless I'm mistaken, it seems like setting the job's duration to the instance duration in `build()`  isn't doing anything, and that code should be removed to avoid confusion.

Ideally, we'd like to get this and the follow on PR #1105 fixed before the tag tomorrow. Maybe I'll look into why the duration adjustment in `run()` isn't working (as best I can), but could use help if you're available.

Reply:
Well, when I print some debug from that line, `graph_end` seems to have the wrong scale:
```
at=1699291701 meta.duration=120 graph_end=1699291819000000000
```
Does `graph_end` need one of those crazy `std::chrono::duration_cast<std::chrono::seconds>`?